File: kafkalite/src/api/message.ts
import { Channel, invoke } from "@tauri-apps/api/core";
import type {
  ApiResult,
  ConsumeRequest,
  SessionEvent,
} from "../types";

function unwrap<T>(result: ApiResult<T>): T {
  if (result.ok && result.data !== undefined) {
    return result.data;
  }
  throw new Error(result.error?.message ?? "Unknown error");
}

export async function startConsume(
  request: ConsumeRequest,
  onEvent: (event: SessionEvent) => void,
): Promise<string> {
  const channel = new Channel<SessionEvent>();
  channel.onmessage = onEvent;
  const result = unwrap(
    await invoke<ApiResult<{ sessionId: string }>>("start_consume", {
      request,
      onEvent: channel,
    }),
  );
  return result.sessionId;
}

export async function stopConsume(sessionId: string): Promise<void> {
  return unwrap(
    await invoke<ApiResult<void>>("stop_consume", { sessionId }),
  );
}

export async function produceTombstone(request: {
  connectionId: string;
  topic: string;
  key: string;
  partition?: number;
}): Promise<{ topic: string; partition: number; offset: number }> {
  return unwrap(
    await invoke<ApiResult<{ topic: string; partition: number; offset: number }>>(
      "produce_tombstone",
      { request },
    ),
  );
}

File: kafkalite/src-tauri/src/commands/message.rs
use serde::Serialize;
use tauri::ipc::Channel;
use tauri::{AppHandle, Manager};
use uuid::Uuid;

use crate::error::{wrap_async, ApiResult};
use crate::kafka::connection_manager::create_producer;
use crate::kafka::consumer_session::{ConsumeRequest, SessionEvent};
use crate::kafka::producer_service::{self, ProduceResult, ProduceTombstoneRequest};
use crate::state::AppState;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StartConsumeResult {
    pub session_id: String,
}

#[tauri::command]
pub async fn start_consume(
    app: AppHandle,
    request: ConsumeRequest,
    on_event: Channel<SessionEvent>,
) -> ApiResult<StartConsumeResult> {
    let operation_id = Uuid::new_v4().to_string();
    let state = app.state::<AppState>();
    wrap_async(operation_id, || async move {
        let profile = {
            let manager = state.connections.lock().map_err(|_| {
                crate::error::AppError::Storage("Connection manager lock poisoned".into())
            })?;
            manager.get_profile(&request.connection_id)?
        };
        let session_id = state.sessions.start(profile, request, on_event).await?;
        Ok(StartConsumeResult { session_id })
    })
    .await
}

#[tauri::command]
pub async fn stop_consume(app: AppHandle, session_id: String) -> ApiResult<()> {
    let operation_id = Uuid::new_v4().to_string();
    let state = app.state::<AppState>();
    wrap_async(operation_id, || async move {
        state.sessions.stop(&session_id).await
    })
    .await
}

#[tauri::command]
pub async fn produce_tombstone(
    app: AppHandle,
    request: ProduceTombstoneRequest,
) -> ApiResult<ProduceResult> {
    let operation_id = Uuid::new_v4().to_string();
    let state = app.state::<AppState>();
    wrap_async(operation_id, || async move {
        let profile = {
            let manager = state.connections.lock().map_err(|_| {
                crate::error::AppError::Storage("Connection manager lock poisoned".into())
            })?;
            manager.get_profile(&request.connection_id)?
        };
        let producer = create_producer(&profile)?;
        let result = producer_service::produce_tombstone(&producer, &request).await?;
        if let Ok(db) = state.db.lock() {
            let _ = db.append_audit(
                Some(&request.connection_id),
                "produce_tombstone",
                Some(&request.topic),
                "ok",
                Some(&request.key),
            );
        }
        Ok(result)
    })
    .await
}
